dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/data/upload/DataOkHttpUploader.kt

```diff
@@ -29,16 +31,22 @@ internal class DataOkHttpUploader(
     val androidInfoProvider: AndroidInfoProvider
 ) : DataUploader {
 
+    private var attempts = 1
+    private var previousUploadStatus: UploadStatus? = null
+    private var previousUploadedBatchId: BatchId? = null
```

Comment:
I particularly looked into this myself also. Initially my intention was to make them thread safe but then I didn't want to add this if it's not really needed. I checked the executors and seen the 1 pool thread size and assumed it is fine. I am thinking that if we really have this scenario that you are mentioning we might end up in different problems as the upload runnable assumes that everything is executed sequentially.

Reply:
> if we really have this scenario that you are mentioning we might end up in different problems as the upload runnable assumes that everything is executed sequentially

And it will be executed sequentially (in the boundaries of the particular feature) even if there are multiple threads in the pool, because there is only one single runnable at the time for the particular feature and the next runnable is scheduled **only** at the end of the execution of the previous feature. And underlying storage is thread-safe, it is designed for the concurrent usage.

Here my point is only about volatile nature of these references. Imagine there is an executor running and thread was killed because of some reason; then new thread will be spinned off which may not see the changes made to these references by the previous thread. Or in the case if executor has multiple threads running.

Reply:
That might be a problem indeed. I think marking them as volatile might make sense to make sure they are being kept in the HEAP and not in the Thread local memory. In case of atomicity we don't need it in this case as we are already running sequential. I am going to open a PR.